llvm/lib/CodeGen/SelectionDAG/InstrEmitter.cpp

```diff
@@ -1193,7 +1193,9 @@ EmitMachineNode(SDNode *Node, bool IsClone, bool IsCloned,
 
   // Run post-isel target hook to adjust this instruction if needed.
   if (II.hasPostISelHook())
-    TLI->AdjustInstrPostInstrSelection(*MIB, Node);
+    TLI->AdjustInstrPostInstrSelection(
+        *MIB, Node,
+        [this, &VRBaseMap](SDValue Op) { return getVR(Op, VRBaseMap); });
```

Comment:
Can you avoid this new callback parameter by adding a map for the tokens to FunctionLoweringInfo?

Reply:
Does that mean that I should pass FLI as a parameter here instead of the lambda? Because TLI does not have a pointer to FLI ...

Reply:
Actually, the `InstEmitter::getVR()` method invoked inside that callback does extra work like creating a new reg for every use of an implicit-def. So the current design looks like the right way to me, unless I am missing something.

Reply:
Yes, I mean pass around FunctionLoweringInfo. 

The implicit_def special casing looks weird to me. Seems like an unnecessary pre-optimization for something that's easy to rematerialize

Reply:
The implicit_def special case shouldn't be relevant here. It should be illegal to have IMPLICIT_DEF define a token

Reply:
Do we really need the target implementation to deal with this? Can we make the assumption that a token operand is the final MI operand and just have InstrEmitter directly handle it?

Reply:
Actually, the reason I went down this path is that `CONVERGENCECTRL_GLUE` is an AMDGPU opcode. I had started by thinking that glueing the token operand to an `SDNode` is a target-specific hack. But at this point, I am fairly confident that this scheme is suitable for all targets. That will simplify other places too, such as the callback passed to `AMDGPUTargetLowering::CollectTargetIntrinsicOperands()`.

As far as I can see, the token still has to be glued, and it eventually becomes an implicit use in the MIR. Otherwise a lot of convergent operations will need to be updated in the TD files.